Test/BlockImpTest.m

```diff
@@ -26,6 +26,7 @@ int main(void)
 {
 	__block int b = 0;
 	void* blk = ^(id self, int a) {
+		assert([self class] == [Foo class]);
```

Comment:
Please can you put the result of `[foo class]` in a local variable before you create the block and then just compare `self` against that global (if `self` is incorrect then we will end up with a crash in `objc_msgSend` for the `[self class]` message, rather than a helpful assert that `self` is wrong.

Reply:
I just tried doing this, and `assert(self == globalFooClass)` always fails. Am I doing the comparison incorrectly?

Reply:
Should work if `globalFooClass` is initialised before the block is created.  What are the two values?  I would simply move line 36 earlier and use that, because that's the value that we're explicitly passing to the first `objc_msgSend` call.

Reply:
They both have the same memory address when I printf them e.g.
```
fooClass: 8081d35c
self: 8081d35c
```
but the `assert(self == fooClass)` doesn't like it.

Reply:
Can you post the complete diff?

Reply:
```
diff --git a/Test/BlockImpTest.m b/Test/BlockImpTest.m
index ce9931a..7fea7ad 100644
--- a/Test/BlockImpTest.m
+++ b/Test/BlockImpTest.m
@@ -24,8 +24,11 @@ +(struct big)sret;
 
 int main(void)
 {
+	Class cls = objc_getClass("Foo");
+	
 	__block int b = 0;
 	void* blk = ^(id self, int a) {
+		assert(self == cls);
 		b += a; 
 		return b; };
 	blk = Block_copy(blk);
@@ -33,7 +36,6 @@ int main(void)
 	char *type = block_copyIMPTypeEncoding_np(blk);
 	assert(NULL != type);
 	class_addMethod((objc_getMetaClass("Foo")), @selector(count:), imp, type);
-	Class cls = objc_getClass("Foo");
 	assert(2 == ((int(*)(id,SEL,int))imp)(cls, @selector(count:), 2));
 	free(type);
 	assert(4 == [Foo count: 2]);
@@ -45,6 +47,7 @@ int main(void)
 	assert(imp_getBlock(imp) != (blk));
 
 	blk = ^(id self) {
+		assert(self == cls);
 		struct big b = {1, 2, 3, 4, 5};
 		return b;
 	};
```

Reply:
Does this fail in the first or second assert?  If you print both values in the block before the assert, what does it show?

Reply:
It fails on the first assert.

If I printf before the first assert:
```
self: 7fcd135c cls: 0
```
So I guess the `0` is the problem.

Reply:
So if I declare

Reply:
That sounds as if there may be a bug in the block copy, but for me clang emits a copy helper with a call to `_Block_object_assign` with and without the `__block` qualifier.

Reply:
What is the best way to check this on my side?

Reply:
Can you put a breakpoint on `_Block_object_assign` and check that, in the `imp_implementationWithBlock` call it's being called for `cls`?  It's also possible that the block trampoline is not actually loading the block into the first argument register correctly.  Do you see the same failure with the ARM and Thumb versions?  The load to `\arg0` is supposed to be from exactly 4 bytes before the block pointer that's called.  It's difficult to check this directly, but you can create a block structure directly using the struct from the `blocks_private.h` header and give it a function pointer that checks the address of the block variable.  We should probably have a test that does that anyway...